File: connectors/connector-redmine/src/main/java/com/taskadapter/connector/redmine/RedmineConnector.java
package com.taskadapter.connector.redmine;

import com.taskadapter.connector.FieldRow;
import com.taskadapter.connector.NewConnector;
import com.taskadapter.connector.common.TaskSavingUtils;
import com.taskadapter.connector.definition.ProgressMonitor;
import com.taskadapter.connector.definition.SaveResult;
import com.taskadapter.connector.definition.TaskId;
import com.taskadapter.connector.definition.WebConnectorSetup;
import com.taskadapter.core.PreviouslyCreatedTasksResolver;
import com.taskadapter.model.AllFields;
import com.taskadapter.model.GTask;
import com.taskadapter.redmineapi.Include;
import com.taskadapter.redmineapi.RedmineException;
import com.taskadapter.redmineapi.RedmineManager;
import com.taskadapter.redmineapi.bean.Issue;
import com.taskadapter.redmineapi.bean.IssuePriority;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class RedmineConnector implements NewConnector {

    /**
     * Keep it the same to enable backward compatibility for previously created config files.
     */
    public static final String ID = "Redmine";

    private final RedmineConfig config;
    private final WebConnectorSetup setup;

    public RedmineConnector(RedmineConfig config, WebConnectorSetup setup) {
        this.config = config;
        this.setup = setup;
    }

    @Override
    public GTask loadTaskByKey(TaskId id, Iterable<FieldRow<?>> rows) {
        var httpClient = RedmineManagerFactory.createRedmineHttpClient(setup.getHost());
        try {
            var mgr = RedmineManagerFactory.createRedmineManager(setup, httpClient);
            var intKey = id.getId().intValue();
            var issue = mgr.getIssueManager().getIssueById(intKey, Include.relations);
            var userCache = loadUsersIfAllowed(mgr);
            var converter = new RedmineToGTask(config, userCache);
            return converter.convertToGenericTask(issue);
        } catch (RedmineException e) {
            throw new RuntimeException(e);
        } finally {
            httpClient.getConnectionManager().shutdown();
        }
    }

    @Override
    public List<GTask> loadData() {
        var httpClient = RedmineManagerFactory.createRedmineHttpClient(setup.getHost());
        try {
            var mgr = RedmineManagerFactory.createRedmineManager(setup, httpClient);
            var usersCache = loadUsersIfAllowed(mgr);
            Integer queryId = config.getQueryId() == null ? null : config.getQueryId().intValue();
            var issues = mgr.getIssueManager().getIssues(config.getProjectKey(), queryId, Include.relations);
            return convertToGenericTasks(config, issues, usersCache);
        } catch (RedmineException e) {
            throw new RuntimeException(e);
        } finally {
            httpClient.getConnectionManager().shutdown();
        }
    }

    private List<GTask> convertToGenericTasks(RedmineConfig config, List<Issue> issues, RedmineUserCache usersCache) {
        var converter = new RedmineToGTask(config, usersCache);
        return issues.stream()
                .map(converter::convertToGenericTask)
                .collect(Collectors.toList());
    }

    @Override
    public SaveResult saveData(PreviouslyCreatedTasksResolver previouslyCreatedTasks, List<GTask> tasks,
                               ProgressMonitor monitor,
                               Iterable<FieldRow<?>> fieldRows) {
        try {
            var httpClient = RedmineManagerFactory.createRedmineHttpClient(setup.getHost());
            var mgr = RedmineManagerFactory.createRedmineManager(setup, httpClient);
            try {
                var rmProject = mgr.getProjectManager().getProjectByKey(config.getProjectKey());
                var priorities = RedmineConnector.loadPriorities(fieldRows, mgr);
                var statusList = mgr.getIssueManager().getStatuses();
                var versions = mgr.getProjectManager().getVersions(rmProject.getId());
                var categories = mgr.getIssueManager().getCategories(rmProject.getId());
                var customFieldDefinitions = mgr.getCustomFieldManager().getCustomFieldDefinitions();
                var userCache = loadUsersIfAllowed(mgr);
                var converter = new GTaskToRedmine(config, priorities, rmProject, userCache, customFieldDefinitions, statusList,
                        versions, categories);
                var saver = new RedmineTaskSaver(mgr.getTransport(), config);
                var tsrb = TaskSavingUtils.saveTasks(previouslyCreatedTasks, tasks, converter, saver, monitor, fieldRows,
                        setup.getHost());
                TaskSavingUtils.saveRemappedRelations(config, tasks, saver, tsrb);
                return tsrb.getResult();
            } finally {
                httpClient.getConnectionManager().shutdown();
            }
        } catch (RedmineException e) {
            return SaveResult.withError(e);
        }
    }

    private RedmineUserCache loadUsersIfAllowed(RedmineManager mgr) throws RedmineException {
        if (!config.isFindUserByName()) {
            return new RedmineUserCache(List.of());
        }
        return new RedmineUserCache(mgr.getUserManager().getUsers());

    }

    private static Map<String, Integer> loadPriorities(Iterable<FieldRow<?>> rows,
                                                       RedmineManager mgr) throws RedmineException {
        if (FieldRowFinder.containsTargetField(rows, AllFields.priority)) {
            return loadPriorities(mgr);
        }
        return new HashMap<>();
    }

    private static Map<String, Integer> loadPriorities(RedmineManager mgr) throws RedmineException {
        return mgr.getIssueManager().getIssuePriorities().stream()
                .collect(Collectors.toMap(IssuePriority::getName, IssuePriority::getId));
    }
}

File: connectors/connector-redmine/src/main/scala/com/taskadapter/connector/redmine/RedmineConnector.scala
package com.taskadapter.connector.redmine

import java.util

import com.taskadapter.connector.{FieldRow, NewConnector}
import com.taskadapter.connector.common.TaskSavingUtils
import com.taskadapter.connector.definition._
import com.taskadapter.core.PreviouslyCreatedTasksResolver
import com.taskadapter.model.{GTask, Priority}
import com.taskadapter.redmineapi.{Include, RedmineException, RedmineManager}
import com.taskadapter.redmineapi.bean._

import scala.collection.JavaConverters._

object RedmineConnector {
  /**
    * Keep it the same to enable backward compatibility for previously created config files.
    */
  val ID = "Redmine"

  @throws[RedmineException]
  private def loadPriorities(rows: java.lang.Iterable[FieldRow[_]], mgr: RedmineManager): util.Map[String, Integer] = {
    if (FieldRowFinder.containsTargetField(rows.asScala.toSeq, Priority))
      loadPriorities(mgr)
    else
      new util.HashMap[String, Integer]
  }

  @throws[RedmineException]
  private def loadPriorities(mgr: RedmineManager): util.Map[String, Integer] = {
    mgr.getIssueManager.getIssuePriorities.asScala.map(p => (p.getName, p.getId)).toMap.asJava
  }
}

class RedmineConnector(config: RedmineConfig, setup: WebConnectorSetup) extends NewConnector {
  override def loadTaskByKey(id: TaskId, rows: Iterable[FieldRow[_]]): GTask = {
    val httpClient = RedmineManagerFactory.createRedmineHttpClient(setup.host)
    try {
      val mgr = RedmineManagerFactory.createRedmineManager(setup, httpClient)
      val intKey = id.id.toInt
      val issue = mgr.getIssueManager.getIssueById(intKey, Include.relations)
      val userCache = loadUsersIfAllowed(mgr)
      val converter = new RedmineToGTask(config, userCache)
      converter.convertToGenericTask(issue)
    } catch {
      case e: RedmineException =>
        throw new RuntimeException(e)
    } finally httpClient.getConnectionManager.shutdown()
  }

  override def loadData(): util.List[GTask] = {
    val httpClient = RedmineManagerFactory.createRedmineHttpClient(setup.host)
    try {
      val mgr = RedmineManagerFactory.createRedmineManager(setup, httpClient)
      val usersCache = loadUsersIfAllowed(mgr)
      val queryId : Integer = if (config.getQueryId == null) {
        null
      } else {
        config.getQueryId.intValue()
      }
      val issues = mgr.getIssueManager.getIssues(config.getProjectKey, queryId, Include.relations)
      convertToGenericTasks(config, issues, usersCache)
    } catch {
      case e: RedmineException =>
        throw new RuntimeException(e)
    } finally httpClient.getConnectionManager.shutdown()
  }

  private def convertToGenericTasks(config: RedmineConfig, issues: util.List[Issue], usersCache: RedmineUserCache) = {
    val converter = new RedmineToGTask(config, usersCache)
    issues.asScala.map(i => converter.convertToGenericTask(i)).asJava
  }

  override def saveData(previouslyCreatedTasks: PreviouslyCreatedTasksResolver, tasks: util.List[GTask],
                        monitor: ProgressMonitor,
                        fieldRows: Iterable[FieldRow[_]]): SaveResult = try {
    val httpClient = RedmineManagerFactory.createRedmineHttpClient(setup.host)
    val mgr = RedmineManagerFactory.createRedmineManager(setup, httpClient)
    try {
      val rmProject = mgr.getProjectManager.getProjectByKey(config.getProjectKey)
      val priorities = RedmineConnector.loadPriorities(fieldRows.asJava, mgr)
      val statusList = mgr.getIssueManager.getStatuses
      val versions = mgr.getProjectManager.getVersions(rmProject.getId)
      val categories = mgr.getIssueManager.getCategories(rmProject.getId)
      val customFieldDefinitions = mgr.getCustomFieldManager.getCustomFieldDefinitions
      val userCache = loadUsersIfAllowed(mgr)
      val converter = new GTaskToRedmine(config, priorities, rmProject, userCache, customFieldDefinitions, statusList,
        versions, categories)
      val saver = new RedmineTaskSaver(mgr.getIssueManager, config)
      val tsrb = TaskSavingUtils.saveTasks(previouslyCreatedTasks, tasks, converter, saver, monitor, fieldRows,
        setup.host)
      TaskSavingUtils.saveRemappedRelations(config, tasks, saver, tsrb)
      tsrb.getResult
    } finally httpClient.getConnectionManager.shutdown()
  } catch {
    case e: RedmineException => throw RedmineExceptions.convertException(e)
  }

  private def loadUsersIfAllowed(mgr: RedmineManager): RedmineUserCache = {
    if (!config.isFindUserByName) {
      new RedmineUserCache(Seq())
    } else {
      new RedmineUserCache(mgr.getUserManager.getUsers.asScala)
    }
  }
}
